metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java

```diff
@@ -171,74 +172,125 @@ private void extractBroaderResources(EnrichmentBase resource, Set<String> destin
+
+    // Parse the entity.
+    final Pair<EnrichmentBase, Vocabulary> enrichmentBaseVocabularyPair;
+    if (entityVocabularyPair == null) {
+      enrichmentBaseVocabularyPair = null;
+    } else {
+      enrichmentBaseVocabularyPair = convertToEnrichmentBaseVocabularyPair(
+          entityVocabularyPair.getLeft(), entityVocabularyPair.getRight());
+    }
+    return enrichmentBaseVocabularyPair;
+  }
+
+  /**
+   * Based on the values of {@code transformedEntityXml} and {@code entityVocabulary} compute the
+   * pair with the transformed entity and vocabulary.
+   * <p>The possible outcomes are:
+   * <ul>
+   *   <li>If the {@code entityVocabulary} is not null then simply return a pair with the
+   *   provided values. We accept that the returned entity can be null, as a result of a
+   *   previous transformation.</li>
+   *   <li>If {@code entityVocabulary} is null then retrieve the original entity and transform it
+   *   based on the vocabulary candidates from the database, and return that pair</li>
+   * </ul>
+   * </p>
+   *
+   * @param resourceId the url of the provider entity
+   * @param cachedEntity the cached entity object
+   * @param transformedEntityXml the previously transformed provider entity
+   * @param entityVocabulary the previously used vocabulary for the transformation
+   * @return a pair with the computed values
+   * @throws URISyntaxException if the resource identifier url is invalid
+   * @throws TransformerException if an exception occurred during transformation of the original
+   * entity
+   */
+  private Pair<String, Vocabulary> computeEntityAndVocabulary(String resourceId,
+      ProcessedEntity cachedEntity, String transformedEntityXml, Vocabulary entityVocabulary)
+      throws URISyntaxException, TransformerException {
+
+    final Pair<String, Vocabulary> transformedEntityAndVocabularyPair;
+    // If the vocabulary is null then we need to find the correct one and transform the provider
+    // entity.
+    if (entityVocabulary == null) {
+      final VocabularyCandidates vocabularyCandidates = VocabularyCandidates
+          .findVocabulariesForUrl(resourceId, vocabularyDao::getByUriSearch);
+      //If any of transformedEntityXml or entityVocabulary is null, we need to re-transform
+      transformedEntityAndVocabularyPair = retrieveEntityAndTransform(resourceId,
+          vocabularyCandidates);
+      //If we have proper results store them
+      if (transformedEntityAndVocabularyPair != null) {
+        //Save entity
+        ProcessedEntity entityToCache =
+            (cachedEntity == null) ? new ProcessedEntity() : cachedEntity;
         entityToCache.setResourceId(resourceId);
-        entityToCache.setVocabularyId((vocabulary == null) ? null : vocabulary.getId().toString());
+        entityToCache.setXml(transformedEntityAndVocabularyPair.getLeft());
+        entityToCache
+            .setVocabularyId(transformedEntityAndVocabularyPair.getRight().getId().toString());
```

Comment:
According to my IDE, getId() is already a string, so no need to call toString() on it.

Reply:
But it's not, did your IDE mess up?

Reply:
Yes it did! :)